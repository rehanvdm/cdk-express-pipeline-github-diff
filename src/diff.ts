import { ResourceDifference, type TemplateDiff } from '@aws-cdk/cloudformation-diff';
import * as fs from 'node:fs';
import { CdkExpressPipelineAssembly } from 'cdk-express-pipeline';

export type DiffOptions = {
  /**
   * The stack selectors to use for the diff operation.
   */
  stackSelectors: string[];
};
export type DiffResult = {
  stacks: Record<string, StackDiff>;
};
export type StackDiff = {
  summary: {
    additions: number;
    removals: number;
    updates: number;
  };
  markdown: string;
};

export function generateDiffs(templateDiffs: { [name: string]: TemplateDiff }): DiffResult | undefined {
  if (Object.keys(templateDiffs).length === 0) {
    return undefined;
  }
  const result: DiffResult = { stacks: {} };
  for (const [name, templateDiff] of Object.entries(templateDiffs)) {
    result.stacks[name] = generateStackDiff(templateDiff);
  }

  return result;
}

export function saveDiffs(diffResult: DiffResult, outputDir: string): void {
  if (Object.keys(diffResult.stacks).length === 0) {
    return;
  }
  for (const [stackNameId, stackDiff] of Object.entries(diffResult.stacks)) {
    if (!fs.existsSync(`${outputDir}/cdk-express-pipeline/diffs`)) {
      fs.mkdirSync(`${outputDir}/cdk-express-pipeline/diffs`, { recursive: true });
    }
    const filePath = `${outputDir}/cdk-express-pipeline/diffs/${stackNameId}.json`;
    fs.writeFileSync(filePath, JSON.stringify(stackDiff, null, 2));
  }
}

export function getSavedDiffs(outputDir: string) {
  const combinedDiff: DiffResult = { stacks: {} };
  const files = fs.readdirSync(`${outputDir}/cdk-express-pipeline/diffs`);
  for (const file of files) {
    const stackId = file.replace('.json', '');
    const stackDiff = JSON.parse(fs.readFileSync(`${outputDir}/cdk-express-pipeline/diffs/${file}`, 'utf-8'));
    combinedDiff.stacks[stackId] = stackDiff;
  }
  return combinedDiff;
}

export function generateMarkdown(order: CdkExpressPipelineAssembly, diffResult: DiffResult) {
  let markdown = '```diff\n';

  order.waves.forEach((wave) => {
    markdown += `🌊 ${wave.waveId}\n`;
    wave.stages.forEach((stage) => {
      markdown += `  🏗 ${stage.stageId}\n`;
      stage.stacks.forEach((stack) => {
        const stackDiff = diffResult.stacks[stack.stackId]; // + ' (' + stack.stackName + ')'
        if (stackDiff) {
          markdown += `    📦 ${stack.stackName} (${stack.stackId})\n`;
          if (stackDiff.markdown) {
            markdown += `${stackDiff.markdown}\n`;
          }
        }
      });
    });
  });
  markdown += '```\n';

  return markdown;
}

function generateStackDiff(templateDiff: TemplateDiff): StackDiff {
  const stackDiff: StackDiff = {
    summary: {
      additions: 0,
      removals: 0,
      updates: 0
    },
    markdown: ''
  };
  const changes: string[] = [];

  templateDiff.resources.forEachDifference((logicalId: string, change: ResourceDifference) => {
    if (ignoreResource(change)) {
      return;
    }

    if (change.isUpdate) {
      stackDiff.summary.updates++;
      const replacementIndicator = change.changeImpact === 'WILL_REPLACE' ? ' [💥 REPLACEMENT]' : '';
      changes.push(
        `!       [~] ${change.oldValue?.Type || change.newValue?.Type} ${logicalId} ${logicalId}${replacementIndicator}`
      );

      Object.entries(change.propertyUpdates).forEach(([propertyPath, propertyChange]) => {
        if (propertyChange.isAddition) {
          changes.push(`!         └─ [+] ${propertyPath}`);
          changes.push(`!             └─ [+] ${JSON.stringify(propertyChange.newValue)}`);
        } else if (propertyChange.isRemoval) {
          changes.push(`!         └─ [-] ${propertyPath}`);
          changes.push(`!             └─ [-] ${JSON.stringify(propertyChange.oldValue)}`);
        } else if (propertyChange.isUpdate) {
          changes.push(`!         └─ [~] ${propertyPath}`);
          changes.push(`!             ├─ [-] ${JSON.stringify(propertyChange.oldValue)}`);
          changes.push(`!             └─ [+] ${JSON.stringify(propertyChange.newValue)}`);
        }
      });
    } else if (change.isAddition) {
      stackDiff.summary.additions++;
      changes.push(`+       [+] ${change.newValue?.Type} ${logicalId} ${logicalId}`);
    } else if (change.isRemoval) {
      stackDiff.summary.removals++;
      changes.push(`-       [-] ${change.oldValue?.Type} ${logicalId} ${logicalId}`);
    }
  });

  if (changes.length > 0) {
    stackDiff.markdown = changes.join('\n');
  }
  return stackDiff;
}

//TODO: Let pass in more later
function ignoreResource(change: ResourceDifference): boolean {
  const resourceType = change.oldValue?.Type ?? change.newValue?.Type;
  switch (resourceType) {
    case 'AWS::CDK::Metadata':
      return true;
    case 'AWS::Lambda::Function': {
      const keys = Object.keys(change.propertyUpdates);
      if ((keys.length <= 2 && keys.includes('Code')) || keys.includes('Metadata')) {
        return true;
      }
    }
  }
  return false;
}
