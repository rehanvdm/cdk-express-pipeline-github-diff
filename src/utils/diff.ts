import { ResourceDifference, type TemplateDiff } from '@aws-cdk/cloudformation-diff';
import * as fs from 'node:fs';
import { CdkExpressPipelineAssembly } from 'cdk-express-pipeline';

export type DiffResult = {
  stacks: Record<string, StackDiff>;
};
export type DiffSummary = {
  additions: number;
  removals: number;
  updates: number;
};
export type StackDiff = {
  summary: DiffSummary;
  markdown: string;
};

export function generateDiffs(templateDiffs: { [name: string]: TemplateDiff }, cdkDiffOutput: string) {
  if (Object.keys(templateDiffs).length === 0) {
    return undefined;
  }
  const result: DiffResult = { stacks: {} };
  for (const [stackIdName, templateDiff] of Object.entries(templateDiffs)) {
    const stackId = stackIdName.split(' ')[0];
    result.stacks[stackId] = generateStackDiff(stackIdName, templateDiff, cdkDiffOutput);
  }

  return result;
}

export function getDiffsDir(outputDir: string) {
  return `${outputDir}/cdk-express-pipeline/diffs`;
}
export function saveDiffs(diffResult: DiffResult, outputDir: string) {
  if (Object.keys(diffResult.stacks).length === 0) {
    return;
  }
  const diffsDir = getDiffsDir(outputDir);
  for (const [stackNameId, stackDiff] of Object.entries(diffResult.stacks)) {
    if (!fs.existsSync(diffsDir)) {
      fs.mkdirSync(diffsDir, { recursive: true });
    }
    const filePath = `${diffsDir}/${stackNameId}.json`;
    fs.writeFileSync(filePath, JSON.stringify(stackDiff, null, 2));
  }
}

export function getSavedDiffs(outputDir: string) {
  const combinedDiff: DiffResult = { stacks: {} };
  const diffsDir = getDiffsDir(outputDir);
  const files = fs.readdirSync(diffsDir);
  for (const file of files) {
    const stackId = file.replace('.json', '');
    const stackDiff = JSON.parse(fs.readFileSync(`${diffsDir}/${file}`, 'utf-8'));
    combinedDiff.stacks[stackId] = stackDiff;
  }
  return combinedDiff;
}

export function generateMarkdown(order: CdkExpressPipelineAssembly, diffResult: DiffResult) {
  let markdown = '```diff\n';

  for (const wave of order.waves) {
    markdown += `🌊 ${wave.waveId}\n`;
    for (const stage of wave.stages) {
      markdown += `  🏗 ${stage.stageId}\n`;
      for (const stack of stage.stacks) {
        const stackDiff = diffResult.stacks[stack.stackId];
        if (stackDiff) {
          markdown += `    📦 ${stack.stackName} (${stack.stackId})\n`;
          if (stackDiff.markdown) {
            markdown += `${stackDiff.markdown}\n`;
          }
        }
      }
    }
  }
  markdown += '```\n';

  return markdown;
}

function generateStackDiff(stackIdName: string, templateDiff: TemplateDiff, cdkDiffOutput: string): StackDiff {
  const stackDiff: StackDiff = {
    summary: {
      additions: 0,
      removals: 0,
      updates: 0
    },
    markdown: ''
  };
  const changes: string[] = [];

  templateDiff.resources.forEachDifference((logicalId: string, change: ResourceDifference) => {
    if (ignoreResource(change)) {
      return;
    }

    if (change.isUpdate) {
      stackDiff.summary.updates++;

      let indicator = '!';
      let replacementIndicator = '';
      if (change.changeImpact === 'WILL_REPLACE') {
        indicator = '-';
        replacementIndicator = ' (requires replacement)';
        stackDiff.summary.removals++;
      } else if (change.changeImpact === 'MAY_REPLACE') {
        indicator = '-';
        replacementIndicator = ' (may require replacement)';
        stackDiff.summary.removals++;
      }

      changes.push(
        `${indicator}       [~] ${change.oldValue?.Type || change.newValue?.Type} ${logicalId} ${logicalId}${replacementIndicator}`
      );

      const propertyEntries = Object.entries(change.propertyUpdates);
      propertyEntries.forEach(([propertyPath, propertyChange], index) => {
        const isLastProperty = index === propertyEntries.length - 1;

        if (propertyChange.isAddition) {
          changes.push(`!         ${isLastProperty ? '└' : '├'}─ [+] ${propertyPath}`);
          changes.push(`!         ${isLastProperty ? ' ' : '│'}   └─ [+] ${JSON.stringify(propertyChange.newValue)}`);
        } else if (propertyChange.isRemoval) {
          changes.push(`!         ${isLastProperty ? '└' : '├'}─ [-] ${propertyPath}`);
          changes.push(`!         ${isLastProperty ? ' ' : '│'}   └─ [-] ${JSON.stringify(propertyChange.oldValue)}`);
        } else if (propertyChange.isUpdate) {
          let propertyIndicator = '!';
          let propertyReplacementIndicator = '';
          if (propertyChange.changeImpact === 'WILL_REPLACE') {
            propertyIndicator = '-';
            propertyReplacementIndicator = ' (requires replacement)';
          } else if (propertyChange.changeImpact === 'MAY_REPLACE') {
            propertyIndicator = '-';
            propertyReplacementIndicator = ' (may require replacement)';
          }

          changes.push(
            `${propertyIndicator}         ${isLastProperty ? '└' : '├'}─ [~] ${propertyPath}${propertyReplacementIndicator}`
          );

          // Check if both values are objects for deep diff
          if (
            propertyChange.oldValue &&
            propertyChange.newValue &&
            typeof propertyChange.oldValue === 'object' &&
            typeof propertyChange.newValue === 'object'
          ) {
            // Do deep diff
            const deepChanges = deepDiff(
              propertyChange.oldValue,
              propertyChange.newValue,
              propertyPath,
              `!         ${isLastProperty ? ' ' : '│'}   `,
              isLastProperty
            );
            if (deepChanges.length > 0) {
              changes.push(...deepChanges);
            } else {
              // Fallback to simple diff if no deep changes found
              changes.push(
                `!         ${isLastProperty ? ' ' : '│'}   ├─ [-] ${JSON.stringify(propertyChange.oldValue)}`
              );
              changes.push(
                `!         ${isLastProperty ? ' ' : '│'}   └─ [+] ${JSON.stringify(propertyChange.newValue)}`
              );
            }
          } else {
            // Simple diff for non-objects
            changes.push(`!         ${isLastProperty ? ' ' : '│'}   ├─ [-] ${JSON.stringify(propertyChange.oldValue)}`);
            changes.push(`!         ${isLastProperty ? ' ' : '│'}   └─ [+] ${JSON.stringify(propertyChange.newValue)}`);
          }
        }
      });
    } else if (change.isAddition) {
      stackDiff.summary.additions++;
      changes.push(`+       [+] ${change.newValue?.Type} ${logicalId} ${logicalId}`);
    } else if (change.isRemoval) {
      stackDiff.summary.removals++;
      changes.push(`-       [-] ${change.oldValue?.Type} ${logicalId} ${logicalId}`);
    }
  });

  if (changes.length > 0) {
    stackDiff.markdown = changes.join('\n');
  }
  return stackDiff;
}

function deepDiff(
  oldValue: any,
  newValue: any,
  path: string = '',
  baseIndent: string = '',
  isLast: boolean = true
): string[] {
  const changes: string[] = [];

  // If both values are objects and not null/undefined, do deep comparison
  if (
    oldValue &&
    newValue &&
    typeof oldValue === 'object' &&
    typeof newValue === 'object' &&
    !Array.isArray(oldValue) &&
    !Array.isArray(newValue)
  ) {
    const allKeys = new Set([...Object.keys(oldValue), ...Object.keys(newValue)]);
    const changedKeys = Array.from(allKeys).filter((key) => oldValue[key] !== newValue[key]);
    let hasChanges = false;

    changedKeys.forEach((key, index) => {
      const oldProp = oldValue[key];
      const newProp = newValue[key];
      const isLastKey = index === changedKeys.length - 1;

      if (oldProp !== newProp) {
        if (
          oldProp &&
          newProp &&
          typeof oldProp === 'object' &&
          typeof newProp === 'object' &&
          !Array.isArray(oldProp) &&
          !Array.isArray(newProp)
        ) {
          // Recursively diff nested objects
          changes.push(`${baseIndent}${isLastKey ? '└' : '├'}─ [~] ${key}`);
          const nestedChanges = deepDiff(oldProp, newProp, key, `${baseIndent}${isLastKey ? ' ' : '│'}   `, true);
          if (nestedChanges.length > 0) {
            changes.push(...nestedChanges);
            hasChanges = true;
          }
        } else if (oldProp === undefined && newProp !== undefined) {
          // Property was added
          changes.push(`${baseIndent}${isLastKey ? '└' : '├'}─ [+] ${key}`);
          changes.push(`${baseIndent}${isLastKey ? ' ' : '│'}   └─ [+] ${JSON.stringify(newProp)}`);
          hasChanges = true;
        } else if (oldProp !== undefined && newProp === undefined) {
          // Property was removed
          changes.push(`${baseIndent}${isLastKey ? '└' : '├'}─ [-] ${key}`);
          changes.push(`${baseIndent}${isLastKey ? ' ' : '│'}   └─ [-] ${JSON.stringify(oldProp)}`);
          hasChanges = true;
        } else {
          // Simple value change
          changes.push(`${baseIndent}${isLastKey ? '└' : '├'}─ [~] ${key}`);
          changes.push(`${baseIndent}${isLastKey ? ' ' : '│'}   ├─ [-] ${JSON.stringify(oldProp)}`);
          changes.push(`${baseIndent}${isLastKey ? ' ' : '│'}   └─ [+] ${JSON.stringify(newProp)}`);
          hasChanges = true;
        }
      }
    });

    return hasChanges ? changes : [];
  } else if (Array.isArray(oldValue) && Array.isArray(newValue)) {
    // Handle arrays
    const maxLength = Math.max(oldValue.length, newValue.length);
    let hasChanges = false;

    for (let i = 0; i < maxLength; i++) {
      const oldItem = oldValue[i];
      const newItem = newValue[i];
      const isLastItem = i === maxLength - 1;

      if (oldItem !== newItem) {
        if (oldItem === undefined) {
          // Item was added
          changes.push(`${baseIndent}${isLastItem ? '└' : '├'}─ [+] ${path}[${i}]`);
          changes.push(`${baseIndent}${isLastItem ? ' ' : '│'}   └─ [+] ${JSON.stringify(newItem)}`);
          hasChanges = true;
        } else if (newItem === undefined) {
          // Item was removed
          changes.push(`${baseIndent}${isLastItem ? '└' : '├'}─ [-] ${path}[${i}]`);
          changes.push(`${baseIndent}${isLastItem ? ' ' : '│'}   └─ [-] ${JSON.stringify(oldItem)}`);
          hasChanges = true;
        } else if (typeof oldItem === 'object' && typeof newItem === 'object' && oldItem !== null && newItem !== null) {
          // Recursively diff array items that are objects
          changes.push(`${baseIndent}${isLastItem ? '└' : '├'}─ [~] ${path}[${i}]`);
          const nestedChanges = deepDiff(
            oldItem,
            newItem,
            `${path}[${i}]`,
            `${baseIndent}${isLastItem ? ' ' : '│'}   `,
            true
          );
          if (nestedChanges.length > 0) {
            changes.push(...nestedChanges);
            hasChanges = true;
          }
        } else {
          // Simple value change in array
          changes.push(`${baseIndent}${isLastItem ? '└' : '├'}─ [~] ${path}[${i}]`);
          changes.push(`${baseIndent}${isLastItem ? ' ' : '│'}   ├─ [-] ${JSON.stringify(oldItem)}`);
          changes.push(`${baseIndent}${isLastItem ? ' ' : '│'}   └─ [+] ${JSON.stringify(newItem)}`);
          hasChanges = true;
        }
      }
    }

    return hasChanges ? changes : [];
  } else {
    // Simple value change (non-objects or arrays)
    changes.push(`${baseIndent}${isLast ? '└' : '├'}─ [~] ${path}`);
    changes.push(`${baseIndent}${isLast ? ' ' : '│'}   ├─ [-] ${JSON.stringify(oldValue)}`);
    changes.push(`${baseIndent}${isLast ? ' ' : '│'}   └─ [+] ${JSON.stringify(newValue)}`);
    return changes;
  }
}

function ignoreResource(change: ResourceDifference): boolean {
  const resourceType = change.oldValue?.Type ?? change.newValue?.Type;
  switch (resourceType) {
    case 'AWS::CDK::Metadata':
      return true;
  }
  return false;
}
