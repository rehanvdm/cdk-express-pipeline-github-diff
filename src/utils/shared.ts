import { createHash } from 'crypto';

export const CDK_EXPRESS_PIPELINE_JSON_FILE = 'cdk-express-pipeline.json';

export function getCacheKey(stackSelector?: string): string {
  let ret = `cdk-diff-pipeline-${process.env.GITHUB_RUN_ID}-${process.env.GITHUB_RUN_ATTEMPT}-`;
  if (stackSelector) {
    ret += createHash('md5').update(stackSelector).digest('hex');
  }
  return ret;
}
