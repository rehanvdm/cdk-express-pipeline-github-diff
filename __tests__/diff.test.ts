import { TemplateDiff } from '@aws-cdk/cloudformation-diff';
import { generateDiffs, generateMarkdown, getSavedDiffs, saveDiffs } from '../src/utils/diff';
import { DiffMethod, ExpandStackSelection, StackSelectionStrategy, Toolkit } from '@aws-cdk/toolkit-lib';
import * as cdk from 'aws-cdk-lib';
import * as sns from 'aws-cdk-lib/aws-sns';
import { CloudAssembly } from 'aws-cdk-lib/cx-api';
import { CdkExpressPipeline, CdkExpressPipelineAssembly, ExpressStack } from 'cdk-express-pipeline';
import path from 'node:path';
import * as fs from 'node:fs';
import { CDK_EXPRESS_PIPELINE_JSON_FILE } from '../src/utils/shared';
import * as lambda from 'aws-cdk-lib/aws-lambda';

type AssemblyDiff = {
  assembly: CloudAssembly;
};
type AssemblyDiffFuncArgs = {
  withChange?: boolean;
  outputDir?: string;
};

function testAssembly(opts?: AssemblyDiffFuncArgs): AssemblyDiff {
  if (opts?.outputDir) {
    process.env.CDK_OUTDIR = opts?.outputDir;
  }
  const app = new cdk.App({
    outdir: opts?.outputDir
  });
  const expressPipeline = new CdkExpressPipeline();
  const wave1 = expressPipeline.addWave('wave1');
  const wave1stage1 = wave1.addStage('stage1');
  const stackA = new ExpressStack(app, 'stack-a', wave1stage1, {
    stackName: 'StackA'
  });
  const wave1stage2 = wave1.addStage('wave1stage2');
  const stackB = new ExpressStack(app, 'stack-b', wave1stage2, {
    stackName: 'StackB'
  });

  const wave2 = expressPipeline.addWave('wave2');
  const wave2stage1 = wave2.addStage('stage1');
  const stackC = new ExpressStack(app, 'stack-c', wave2stage1, {
    stackName: 'StackC'
  });

  if (!opts?.withChange) {
    new sns.Topic(stackA, 'TopicA', {
      displayName: 'Topic A'
    });
    new sns.Topic(stackB, 'TopicB', {
      displayName: 'Topic B'
    });
    new sns.Topic(stackB, 'TopicR', {
      topicName: 'Topic R',
      displayName: 'Topic R',
      loggingConfigs: [
        {
          protocol: sns.LoggingProtocol.HTTP
        }
      ]
    });

    const vpc = new cdk.aws_ec2.Vpc(stackC, 'VPC');
    const sg1 = new cdk.aws_ec2.SecurityGroup(stackC, 'SG1', {
      vpc
    });
    sg1.addIngressRule(cdk.aws_ec2.Peer.anyIpv4(), cdk.aws_ec2.Port.tcp(443), 'Allow HTTPS traffic');
    new lambda.Function(stackC, 'Function', {
      runtime: lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      code: lambda.Code.fromInline('CHANGED CODE'),
      environment: {
        key1: 'value1',
        key2: 'value2'
      },
      memorySize: 256,
      vpc: vpc,
      securityGroups: [
        sg1,
        new cdk.aws_ec2.SecurityGroup(stackC, 'SG2', {
          vpc
        })
      ]
    });
  } else {
    new sns.Topic(stackA, 'TopicA', {
      displayName: 'Topic A Change'
    });
    new sns.Topic(stackC, 'TopicC', {
      displayName: 'Topic C'
    });
    new sns.Topic(stackB, 'TopicR', {
      topicName: 'Topic R should not change',
      displayName: 'Topic R can change',
      enforceSSL: true
    });

    const vpc = new cdk.aws_ec2.Vpc(stackC, 'VPC', {
      vpcName: 'VPC Changed',
      maxAzs: 2
    });
    const sg1 = new cdk.aws_ec2.SecurityGroup(stackC, 'SG1', {
      vpc
    });
    sg1.addIngressRule(cdk.aws_ec2.Peer.anyIpv4(), cdk.aws_ec2.Port.tcp(443), 'Allow HTTPS traffic');
    new lambda.Function(stackC, 'Function', {
      runtime: lambda.Runtime.NODEJS_18_X,
      handler: 'index.handler',
      code: lambda.Code.fromInline('exports.handler = async function(event, context) { return "Hello World"; };'),
      environment: {
        key1: 'value1-change',
        key3: 'value3'
      },
      memorySize: 512,
      vpc: vpc,
      securityGroups: [sg1]
    });
  }

  expressPipeline.synth([wave1, wave2], false, {});
  process.env.CDK_OUTDIR = undefined;

  return {
    assembly: app.synth()
  };
}

async function generateTemplateDiffs(diffFunc: (opts?: AssemblyDiffFuncArgs) => AssemblyDiff, cdkOutChange: string) {
  let cdkDiffOutput = '';
  const cdkToolkit = new Toolkit({
    color: false,
    ioHost: {
      notify: async function (msg) {
        if (msg.level === 'result') {
          cdkDiffOutput += msg.message + '\n';
        }
      },
      requestResponse: async function (msg) {
        if (msg.level === 'result') {
          cdkDiffOutput += msg.message + '\n';
        }
        return msg.defaultResponse;
      }
    }
  });

  if (fs.existsSync(cdkOutChange)) {
    fs.rmSync(cdkOutChange, { recursive: true, force: true });
  }

  // Synth current CDK to file
  const diffBefore = diffFunc({ outputDir: cdkOutChange });
  // Synth changed CDK and only keep in memory
  const diffAfter = diffFunc({ withChange: true });
  const cxAfter = await cdkToolkit.fromAssemblyBuilder(async () => diffAfter.assembly);

  let templateDiffs: { [id: string]: TemplateDiff } = {};
  for (const stack of diffBefore.assembly.stacks) {
    const templateDiff = await cdkToolkit.diff(cxAfter, {
      method: DiffMethod.LocalFile(path.join(cdkOutChange, stack.id + '.template.json')),
      stacks: {
        strategy: StackSelectionStrategy.PATTERN_MUST_MATCH_SINGLE,
        patterns: [stack.id],
        expand: ExpandStackSelection.NONE,
        failOnEmpty: false
      }
    });
    templateDiffs = {
      ...templateDiffs,
      ...templateDiff
    };
  }
  // console.log('cdkConsole');
  // console.log(cdkConsole);

  return { templateDiffs, cdkDiffOutput };
}

describe('diff.ts', () => {
  it('test complex diff markdown', async () => {
    const cdkOut = path.join(__dirname, 'fixtures', 'cdk.out', 'testAssembly');

    // GH Action 1
    const testDiffRes = await generateTemplateDiffs(testAssembly, cdkOut);
    const stackDiffs = await generateDiffs(testDiffRes.templateDiffs, testDiffRes.cdkDiffOutput);
    if (stackDiffs) {
      await saveDiffs(stackDiffs, cdkOut);
    }

    // GH Action 2
    const allStackDiffs = getSavedDiffs(cdkOut);
    const shortHandOrder: CdkExpressPipelineAssembly = JSON.parse(
      fs.readFileSync(path.join(cdkOut, CDK_EXPRESS_PIPELINE_JSON_FILE), 'utf-8')
    );
    const markdown = generateMarkdown(shortHandOrder, allStackDiffs);

    // Local inspection
    //fs.writeFileSync('__tests__/diff-output-markdown.md', result);

    expect(markdown).toMatchSnapshot();
  });
});
