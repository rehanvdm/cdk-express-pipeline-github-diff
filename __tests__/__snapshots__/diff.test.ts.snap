// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`diff.ts test complex diff markdown 1`] = `
"---
\`\`\`diff
🌊 wave1
  🏗 stage1
    📦 StackA (wave1_stage1_stack-a)
!       [~] AWS::SNS::Topic TopicA1C813746 TopicA1C813746
!         └─ [~] DisplayName
!             ├─ [-] "Topic A"
!             └─ [+] "Topic A Change"
  🏗 wave1stage2
    📦 StackB (wave1_wave1stage2_stack-b)
-       [-] AWS::SNS::Topic TopicB2E0A749C TopicB2E0A749C
🌊 wave2
  🏗 stage1
    📦 StackC (wave2_stage1_stack-c)
+       [+] AWS::SNS::Topic TopicCB48CAADF TopicCB48CAADF
\`\`\`
"
`;
